src/dune_engine/sandbox_mode.ml

```diff
@@ -70,51 +78,43 @@ module Dict = struct
 end
 
 module Set = struct
-  type key = t
-  type t = bool Dict.t
-
-  let compare = Dict.compare Bool.compare
-  let of_func = Dict.of_func
-  let singleton k = of_func (equal k)
+  module T = struct
+    type nonrec t = t
+
+    let to_int = function
+      | None -> 1
+      | Some Copy -> 2
+      | Some Symlink -> 3
+      | Some Hardlink -> 4
+      | Some Patch_back_source_tree -> 5
+    ;;
+
+    let all =
+      [ None; Some Copy; Some Symlink; Some Hardlink; Some Patch_back_source_tree ]
+    ;;
+
+    let to_dyn = to_dyn
+  end
+
+  include Bit_set.Make (T)
+
+  let of_func =
+    let ( ++ ) = union in
+    fun f ->
+      let e x = if f x then singleton x else empty in
+      e None
+      ++ e (Some Copy)
+      ++ e (Some Symlink)
+      ++ e (Some Hardlink)
+      ++ e (Some Patch_back_source_tree)
```

Comment:
I think there should be `Bit_set.of_func`, as this code seems fiddly enough to be hidden behind the API.

Reply:
Sure, although note that the implementation will not be as good. As we'll need to fold over the list of elements instead of just inlining everything.

Reply:
I think you could do `Array.iteri` over a precomputed array of all elements but perhaps that's still less efficient? In any case, it seems to me that this is unlikely to matter since creating a rule involves a lot more work than traversing a little array.

Reply:
Sure. Sounds good.